modules/series-service-impl/src/main/java/org/opencastproject/series/impl/SeriesServiceImpl.java

```diff
@@ -504,53 +504,78 @@ public void repopulate() throws IndexRebuildException {
     try {
       List<SeriesEntity> databaseSeries = persistence.getAllSeries();
       final int total = databaseSeries.size();
-      int current = 1;
       logIndexRebuildBegin(logger, index.getIndexName(), total, "series");
-
-      for (SeriesEntity series: databaseSeries) {
-        Organization organization = orgDirectory.getOrganization(series.getOrganization());
-        User systemUser = SecurityUtil.createSystemUser(systemUserName, organization);
-        SecurityUtil.runAs(securityService, organization, systemUser,
-                () -> {
-                  String seriesId = series.getSeriesId();
-                  logger.trace("Adding series {} for organization {} to the {} index.", seriesId,
-                          series.getOrganization(), index.getIndexName());
-                  List<Function<Optional<Series>, Optional<Series>>> updateFunctions = new ArrayList<>();
-
-                  DublinCoreCatalog catalog;
-                  try {
-                    catalog = DublinCoreXmlFormat.read(series.getDublinCoreXML());
-                    updateFunctions.add(getMetadataUpdateFunction(seriesId, catalog, organization.getId()));
-                  } catch (IOException | ParserConfigurationException | SAXException e) {
-                    logger.error("Could not read dublincore XML of series {}.", seriesId, e);
-                    return;
-                  }
-
-                  String aclStr = series.getAccessControl();
-                  if (StringUtils.isNotBlank(aclStr)) {
-                    try {
-                      AccessControlList acl = AccessControlParser.parseAcl(aclStr);
-                      updateFunctions.add(getAclUpdateFunction(seriesId, acl, organization.getId()));
-                    } catch (Exception ex) {
-                      logger.error("Unable to parse ACL of series {}.", seriesId, ex);
+      if (total > 0) {
+        String lastOrg = databaseSeries.get(0).getOrganization();
+        int current = 0;
+        int n = 16;
+        while (current < total) {
+          List<SeriesEntity> seriesBundle = new ArrayList<>();
+          for (int i = 0; i < n; i++) {
+            if (current + i >= total) {
+              break;
+            }
+            String currentOrg = databaseSeries.get(current + i).getOrganization();
+            if (currentOrg.equals(lastOrg)) {
+              seriesBundle.add(databaseSeries.get(current + i));
+            } else {
+              lastOrg = currentOrg;
+              break;
+            }
```

Comment:
This is actually more about the complex way of passing series along to the actual update code, but this would also solve the organization issue why I put it here. Instead of passing along lists of identifiers and update functions, why not just pass on a list of data to update? That simplifies the code quite a bit and you could have something like:

```java

  var seriesBlock = new ArrayList<Series>();
  for (SeriesEntity series: seriesBundle) {
    String seriesId = series.getSeriesId();
    logger.trace("Adding series {} for organization {} to the {} index.", seriesId,
        series.getOrganization(), index.getIndexName());
    Organization organization = orgDirectory.getOrganization(series.getOrganization());
    User systemUser = SecurityUtil.createSystemUser(systemUserName, organization);
    SecurityUtil.runAs(securityService, organization, systemUser,
          () -> {
              var seriesData = Optional.of(new Series(seriesId, organization.getId()));
              try {
                DublinCoreCatalog catalog = DublinCoreXmlFormat.read(series.getDublinCoreXML());
                seriesData = getMetadataUpdateFunction(seriesId, catalog, organization.getId())
                    .apply(seriesData);
              } catch (IOException | ParserConfigurationException | SAXException e) {
                logger.error("Could not read dublincore XML of series {}.", seriesId, e);
                return;
              }

              String aclStr = series.getAccessControl();
              if (StringUtils.isNotBlank(aclStr)) {
                try {
                  AccessControlList acl = AccessControlParser.parseAcl(aclStr);
                  seriesData = getAclUpdateFunction(seriesId, acl, organization.getId())
                      .apply(seriesData);
                } catch (Exception ex) {
                  logger.error("Unable to parse ACL of series {}.", seriesId, ex);
                }
              }

              try {
                Map<String, String> properties = persistence.getSeriesProperties(seriesId);
                seriesData = getThemePropertyUpdateFunction(seriesId,
                        Optional.ofNullable(properties.get(THEME_PROPERTY_NAME)), organization.getId())
                    .apply(seriesData);
              } catch (NotFoundException | SeriesServiceDatabaseException e) {
                logger.error("Error reading properties of series {}", seriesId, e);
              }
              seriesBlock.add(seriesData.get());
          });
  }

  // do the actual index update
  updateSeriesBundleInIndex(seriesBlock);
```

You could even just run this loop over all data and then call update index if `seriesBlock.size() >= n`.

Reply:
I just modified the original process, so that we can use multiple elements for bulk ingests. But dealing with the updated elements instead of the update functions and all the other stuff will also remove at least two additional abstraction layers. I really would prefer that and I'll try to integrate this approach in the other services, too.